File: frontend/src/types/metadata.ts
export interface MetadataDomainStat {
  business_domain: string;
  table_count: number;
}

export interface MetadataTopTable {
  table_name: string;
  business_domain: string;
  column_count: number;
}

export interface MetadataSummary {
  total_tables: number;
  valid_tables: number;
  offline_tables: number;
  total_columns: number;
  valid_columns: number;
  sensitive_columns: number;
  total_metrics: number;
  valid_metrics: number;
  domain_distribution: MetadataDomainStat[];
  top_table_columns: MetadataTopTable[];
}

export interface MetadataTableListItem {
  id: number;
  table_name: string;
  table_comment?: string;
  business_domain?: string;
  owner?: string;
  update_frequency?: string;
  is_valid: boolean;
  column_count: number;
  sensitive_column_count: number;
  updated_at?: string;
}

export interface MetadataTableList {
  items: MetadataTableListItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface MetadataColumnItem {
  id: number;
  column_name: string;
  ordinal_position?: number;
  data_type?: string;
  column_comment?: string;
  business_desc?: string;
  example_value?: string;
  is_primary_key: boolean;
  is_nullable?: boolean | null;
  is_sensitive: boolean;
  is_valid: boolean;
}

export interface MetadataMetricItem {
  id: number;
  metric_code: string;
  metric_name: string;
  formula?: string;
  business_desc?: string;
  source_tables?: string[];
  source_columns?: string[];
  frequency?: string;
  is_valid: boolean;
}

export interface MetadataTableDetail {
  table: MetadataTableListItem;
  columns: MetadataColumnItem[];
  related_metrics: MetadataMetricItem[];
}

export interface MetadataGraphNode {
  name: string;
  value?: number;
  table_name?: string;
  table_comment?: string;
  table_count?: number;
  is_valid?: boolean;
  children?: MetadataGraphNode[];
}

export interface MetadataGraph {
  tree: MetadataGraphNode;
}

export interface MetadataQualityIssueStat {
  issue_code: string;
  issue_name: string;
  count: number;
}

export interface MetadataQualityDomainStat {
  business_domain: string;
  table_count: number;
  valid_table_count: number;
  avg_quality_score: number;
  unclassified: boolean;
}

export interface MetadataQualityTableItem {
  id: number;
  table_name: string;
  table_comment?: string;
  business_domain?: string;
  is_valid: boolean;
  column_count: number;
  primary_key_count: number;
  missing_column_comment_count: number;
  missing_business_desc_count: number;
  unknown_nullable_count: number;
  lineage_ref_count: number;
  quality_score: number;
  issue_tags: string[];
}

export interface MetadataQualitySummary {
  total_tables: number;
  valid_tables: number;
  offline_tables: number;
  unclassified_tables: number;
  tables_missing_comment: number;
  tables_without_columns: number;
  tables_without_primary_key: number;
  total_columns: number;
  columns_missing_comment: number;
  columns_missing_business_desc: number;
  columns_unknown_nullable: number;
  columns_missing_ordinal_position: number;
  columns_type_anomaly: number;
  table_comment_coverage: number;
  column_comment_coverage: number;
  primary_key_coverage: number;
  issue_stats: MetadataQualityIssueStat[];
  domain_quality: MetadataQualityDomainStat[];
  top_issue_tables: MetadataQualityTableItem[];
}

export interface MetadataTableQuery {
  keyword?: string;
  business_domain?: string;
  is_valid?: boolean;
  sort_by?: string;
  sort_order?: "asc" | "desc";
  page?: number;
  page_size?: number;
}

File: backend/app/schemas/metadata.py
from datetime import datetime

from pydantic import BaseModel


class MetadataDomainStat(BaseModel):
    business_domain: str
    table_count: int


class MetadataTopTable(BaseModel):
    table_name: str
    business_domain: str
    column_count: int


class MetadataSummary(BaseModel):
    total_tables: int
    valid_tables: int
    offline_tables: int
    total_columns: int
    valid_columns: int
    sensitive_columns: int
    total_metrics: int
    valid_metrics: int
    domain_distribution: list[MetadataDomainStat]
    top_table_columns: list[MetadataTopTable]


class MetadataTableListItem(BaseModel):
    id: int
    table_name: str
    table_comment: str | None = None
    business_domain: str | None = None
    owner: str | None = None
    update_frequency: str | None = None
    is_valid: bool
    column_count: int
    sensitive_column_count: int
    updated_at: datetime | None = None


class MetadataTableList(BaseModel):
    items: list[MetadataTableListItem]
    total: int
    page: int
    page_size: int


class MetadataColumnItem(BaseModel):
    id: int
    column_name: str
    ordinal_position: int | None = None
    data_type: str | None = None
    column_comment: str | None = None
    business_desc: str | None = None
    example_value: str | None = None
    is_primary_key: bool = False
    is_nullable: bool | None = None
    is_sensitive: bool
    is_valid: bool


class MetadataMetricItem(BaseModel):
    id: int
    metric_code: str
    metric_name: str
    formula: str | None = None
    business_desc: str | None = None
    source_tables: list[str] | None = None
    source_columns: list[str] | None = None
    frequency: str | None = None
    is_valid: bool


class MetadataTableDetail(BaseModel):
    table: MetadataTableListItem
    columns: list[MetadataColumnItem]
    related_metrics: list[MetadataMetricItem]


class MetadataGraph(BaseModel):
    tree: dict


class MetadataQualityIssueStat(BaseModel):
    issue_code: str
    issue_name: str
    count: int


class MetadataQualityDomainStat(BaseModel):
    business_domain: str
    table_count: int
    valid_table_count: int
    avg_quality_score: float
    unclassified: bool


class MetadataQualityTableItem(BaseModel):
    id: int
    table_name: str
    table_comment: str | None = None
    business_domain: str | None = None
    update_frequency: str | None = None
    is_valid: bool
    column_count: int
    primary_key_count: int
    missing_column_comment_count: int
    missing_business_desc_count: int
    unknown_nullable_count: int
    lineage_ref_count: int
    quality_score: int
    issue_tags: list[str]
    lifecycle_hint: str
    lifecycle_reasons: list[str]
    updated_at: datetime | None = None


class MetadataUnclassifiedDiagnosticStat(BaseModel):
    diagnostic_code: str
    diagnostic_name: str
    count: int
    description: str


class MetadataQualitySummary(BaseModel):
    total_tables: int
    valid_tables: int
    offline_tables: int
    unclassified_tables: int
    tables_missing_comment: int
    tables_without_columns: int
    tables_without_primary_key: int
    total_columns: int
    columns_missing_comment: int
    columns_missing_business_desc: int
    columns_unknown_nullable: int
    columns_missing_ordinal_position: int
    columns_type_anomaly: int
    table_comment_coverage: float
    column_comment_coverage: float
    primary_key_coverage: float
    issue_stats: list[MetadataQualityIssueStat]
    unclassified_diagnostics: list[MetadataUnclassifiedDiagnosticStat]
    domain_quality: list[MetadataQualityDomainStat]
    top_issue_tables: list[MetadataQualityTableItem]
